flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/OuterJoinRecordStateViews.java

```diff
@@ -160,6 +161,17 @@ static void processElements(JoinRecordStateView otherView, Collector<RowData> co
                 collect.collect(outRow);
             }
         }
+
+        if (!isAntiJoin && StreamingPerfOptimizationUtil.isEligibleForNullEquijoinOptimization(
+                StreamingPerfOptimizationUtil.isKeyAnyNulls(thisRow),
+                isEquijoin,
+                false, // We are evicting state and transitioning to streaming so set batch_mode to False
+                statelessNullKeysEnabled,
+                true)) {
+            // Retract record from state if it is an outer equijoin on a null key (Reduces operator state we need to maintain)
+            LOG.debug("Performance Optimization - Evicting NULL record from state in outer equijoin");
+            thisView.retractRecord(thisRow);
```

Comment:
let's add counters here: both for number of records emitted as well as number of records removed from state

Reply:
done